scripts/post_deploy_tdr.py

```diff
@@ -23,78 +34,96 @@
-    }
-    assert tdr_catalogs, tdr_catalogs
-    futures = []
-    all_sources = set()
-    with ThreadPoolExecutor(max_workers=8) as tpe:
-        for catalog in tdr_catalogs:
-            catalog_sources = config.sources(catalog)
-            for source in catalog_sources - all_sources:
-                source = TDRSourceSpec.parse(source)
-                futures.append(tpe.submit(verify_source, catalog, source))
-            all_sources |= catalog_sources
-        for completed_future in as_completed(futures):
-            futures.remove(completed_future)
-            e = completed_future.exception()
-            if e is not None:
-                for running_future in futures:
-                    running_future.cancel()
-                raise e
-
-
-def verify_source(catalog: CatalogName, source_spec: TDRSourceSpec):
-    source = tdr.lookup_source(source_spec)
-    log.info('TDR client is authorized for API access to %s.', source_spec)
-    require(source.project == source_spec.project,
-            'Actual Google project of TDR source differs from configured one',
-            source.project, source_spec.project)
-    # Uppercase is standard for multi-regions in the documentation but TDR
-    # returns 'us' in lowercase
-    require(source.location.lower() == config.tdr_source_location.lower(),
-            'Actual storage location of TDR source differs from configured one',
-            source.location, config.tdr_source_location)
-    # FIXME: Eliminate azul.terra.TDRClient.TDRSource
-    #        https://github.com/DataBiosphere/azul/issues/5524
-    ref = TDRSourceRef(id=source.id, spec=source_spec)
-    plugin = client.repository_plugin(catalog)
-    subgraph_count = sum(plugin.list_partitions(ref).values())
-    require(subgraph_count > 0,
-            'Source spec is empty (bad prefix?)', source_spec)
-
-
-def verify_source_access():
-    public_snapshots = public_tdr.snapshot_ids()
-    all_snapshots = tdr.snapshot_ids()
-    diff = public_snapshots - all_snapshots
-    require(not diff,
-            'The public service account can access snapshots that the indexer '
-            'service account cannot', diff)
+
+class TerraValidator:
+
+    @cached_property
+    def catalogs(self) -> Collection[CatalogName]:
+        result = [
+            catalog.name
+            for catalog in config.catalogs.values()
+            if config.is_tdr_enabled(catalog.name)
+        ]
+        assert result, config.catalogs
+        return result
+
+    @cached_property
+    def azul_client(self) -> AzulClient:
+        return AzulClient()
+
+    def repository_plugin(self, catalog: CatalogName) -> TDRPlugin:
+        assert catalog in self.catalogs, catalog
+        plugin = self.azul_client.repository_plugin(catalog)
+        assert isinstance(plugin, TDRPlugin), plugin
+        return plugin
+
+    @property
+    def tdr(self) -> TDRClient:
+        # This property is cached per-thread on the plugin instance
```

Comment:
> Here, I'd rather not involve the plugin at all if possible.

The plugin is needed to validate that the configured partitions are not empty. 
https://github.com/DataBiosphere/azul/blob/da285b8b2c1e33f0b07d8405bf0f3e40087d53c3/scripts/post_deploy_tdr.py#L79

Reply:
> We don't need per thread caching. Just a per thread instance of the client. This can be achieved by having each worker allocate an instance and discard it when done.

My understanding is that "worker" means "thread" in this context. It is not obvious to me how a `Thread` instance can allocate a TDR client. We could subclass `threading.Thread`, but then we'd have to stop using `ThreadPoolExecutor` as AFAIK it doesn't support alternative thread classes.

Given that each worker thread handles many different function executions (there is 1 future per snapshot--i.e., potentially hundreds--and only 8 threads), we can't unconditionally allocate the client within the submitted function or there will be many clients per thread instead of 1. However, we also can't allocate the client before the function is submitted to the `ThreadPoolExecutor`, because we do not know in advance which thread it will be delegated to.

To summarize, I think that if we do not use per-thread caching, we will need to abandon `ThreadPoolExecutor` in favor of a more flexible and complicated multithreading approach.

Reply:
Agreed at PL that we can use per-thread caching, just don't rely on the plugin for it.